### PR TITLE
enh(charts): add new methods in graph backend

### PR DESCRIPTION
**Pull request by garnier-quentin**
## Description

Add a method in backend + widget preferences. 

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 2.8.x
- [ ] 18.10.x
- [ ] 19.04.x
- [x] 19.10.x
- [x] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [x] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

Refs: MON-4576

### CODE REVIEW

File: www/api/class/centreon_metric.class.php [hunk 1]
```diff
@@ -155,6 +155,44 @@ protected function getListByService()
...
     }
 
+    /**
+     * Get last metrics value
+     *
+     * @return array
+     */
+    public function getLastMetricsData()
+    {
+        if (false === isset($this->arguments['services']) &&
```
Comment (agelibert) [inline www/api/class/centreon_metric.class.php]: Shouldnt compare boolean with boolean to get another boolean, in a boolean evaluated expression :

if (!isset($this->arguments['services']) && !isset($this->arguments['metrics'])

File: www/api/class/centreon_metric.class.php [hunk 2]
```diff
@@ -155,6 +155,44 @@ protected function getListByService()
...
+
+
+    /**
+     * Get metrics datas for a services, metrics
+     *
+     * @return array
+     */
+    public function getMetricsData()
+    {
+        if (false === isset($this->arguments['services']) &&
```
Comment (agelibert) [inline www/api/class/centreon_metric.class.php]: same booleanception there

File: www/api/class/centreon_metric.class.php [hunk 3]
```diff
@@ -214,6 +252,286 @@ public function getMetricsDataByMetric()
...
+    /**
+     * Check acl for user
+     * 
+     * @param int    $hostId     Host id to check
+     * @param int    $serviceId  Service id to check
+     * @param string $aclGroups  String with user acl groups
+     * @param int    $isAdmin    User is admin or not
+     * @param int    $exception  Throw exception or return integer
+     *
+     * @return int
```
Comment (agelibert) [inline www/api/class/centreon_metric.class.php]: maybe better to describe the meaning of returned value there ?

File: www/api/class/centreon_metric.class.php [hunk 4]
```diff
@@ -214,6 +252,286 @@ public function getMetricsDataByMetric()
...
+        $isAdmin = $centreon->user->admin;
+        $aclGroups = null;
+
+        /* Get ACL if user is not admin */
+        if (!$isAdmin) {
+            $acl = new CentreonACL($userId, $isAdmin);
+            $aclGroups = $acl->getAccessGroupsString();
+        }
+
+        if (false === isset($this->arguments['start']) ||
```
Comment (agelibert) [inline www/api/class/centreon_metric.class.php]: booleanception

File: www/api/class/centreon_metric.class.php [hunk 5]
```diff
@@ -214,6 +252,286 @@ public function getMetricsDataByMetric()
...
+
+    /**
+     * Get last data for metrics (by services and/or metrics)
+     *
+     * @param string $services List of services (like hostId_serviceId,hostId2_serviceId2,...)
+     * @param string $metrics  List of metrics (like metricId,metricId2,...)
+     *
+     * @return array
+     *
+     * @throws Exception
```
Comment (agelibert) [inline www/api/class/centreon_metric.class.php]: Is this comment turning unchecked exception to checked exception ?
If not, describe the case that triggers it, or remove the comment.

File: www/class/centreonGraphNg.class.php [hunk 6]
```diff
@@ -1004,10 +1086,23 @@ private function formatByMetrics($rrdData)
...
                 }
             }
-            
+
+            $minimum_value = null;
+            $maximum_value = null;
             for ($j = 0; $j < $size; $j++) {
                 $metric['data'][] = $rrdData['data'][$j][$i];
+                if (!is_null($rrdData['data'][$j][$i]) && 
+                    (is_null($minimum_value) || $rrdData['data'][$j][$i] < $minimum_value)) {
```
Comment (agelibert) [inline www/class/centreonGraphNg.class.php]: `$rrdData['data'][$j][$i]`

feel like we are mixing up the how and the what
+
there is redondancy with the following tests & assignments (violates DRY)

=

assign the rrdData current index in explicit variable telling the what

File: www/api/class/centreon_metric.class.php (hunk 1 above)
Comment (garnier-quentin) [reply www/api/class/centreon_metric.class.php]: i have followed code from centreon php team. Me, i prefer your syntax.

File: www/class/centreonGraphNg.class.php [hunk 7]
```diff
@@ -218,7 +218,9 @@ public function __construct($userId)
         $this->listMetricsId = array();
         $this->metrics = array();
         $this->vmetrics = array();
+        $this->templateInformations = array();
         $this->extraDatas = array();
+        $this->multipleServices = 0;
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
        $this->multipleServices = false;
```

File: www/class/centreonGraphNg.class.php [hunk 8]
```diff
@@ -244,7 +246,19 @@ public function __construct($userId)
...
-    
+
+    /**
+     * Tell it will be a multiple services graph
+     *
+     * @param int $multiple set multiple value
+     *
+     * @return void
+     */
+    public function setMultipleServices($multiple=1)
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
    public function setMultipleServices($multiple)
```

File: www/api/class/centreon_metric.class.php [hunk 9]
```diff
@@ -214,6 +248,294 @@ public function getMetricsDataByMetric()
...
+     * @throws RestBadRequestException
+     * @throws RestForbiddenException
+     * @throws RestNotFoundException
+     */
+    protected function metricsData($services, $metrics)
+    {
+        global $centreon;
+
+        $selectedMetrics = $this->manageMetricsDataArguments($services, $metrics);
+        $multipleServices = count(array_keys($selectedMetrics)) > 1 ? 1 : 0;
```
Comment (sc979) [inline www/api/class/centreon_metric.class.php]:
```suggestion
        $multipleServices = count(array_keys($selectedMetrics)) > 1 ? true : false;
```

File: www/api/class/centreon_metric.class.php [hunk 10]
```diff
@@ -214,6 +248,294 @@ public function getMetricsDataByMetric()
...
+                    $graph->addServiceCustomMetrics($hostId, $serviceId, $metrics);
+                }
+            }
+        } catch (Exception $e) {
+            throw new RestNotFoundException("Graph not found");
+        }
+
+        $result = $graph->getGraph($this->arguments['start'], $this->arguments['end']);
+
+        if ($multipleServices == 0 && count($selectedMetrics) > 0) {
```
Comment (sc979) [inline www/api/class/centreon_metric.class.php]:
```suggestion
        if (!$multipleServices && count($selectedMetrics) > 0) {
```

File: www/api/class/centreon_metric.class.php [hunk 11]
```diff
@@ -673,6 +991,32 @@ protected function convertNaN($element)
...
+     * Get the list of a acknowlegments for a service during a period
+     *
+     * @param int $hostId the host id
+     * @param int $serviceId the service id
+     * @param int $start the start timestamp
+     * @param int $end the end timestamp
+     *
+     * @return array The list of acknowledgements
+     */
+    protected function getAcknowledgements(int $hostId, int $serviceId, int $start, int $end)
```
Comment (sc979) [inline www/api/class/centreon_metric.class.php]:
```suggestion
    protected function getAcknowledgements(int $hostId, int $serviceId, int $start, int $end): array
```

File: www/api/class/centreon_metric.class.php [hunk 12]
```diff
@@ -701,24 +1045,22 @@ protected function getAcknowlegePeriods($hostId, $serviceId, $start, $end)
...
+     * Get list of downtime periods by service
+     *
+     * @param int $hostId the host id
+     * @param int $serviceId the service id
+     * @param int $start the start timestamp
+     * @param int $end the end timestamp
      *
      * @return array
      */
     protected function getDowntimePeriods($hostId, $serviceId, $start, $end)
```
Comment (sc979) [inline www/api/class/centreon_metric.class.php]:
```suggestion
    protected function getDowntimePeriods($hostId, $serviceId, $start, $end): array
```

File: www/class/centreonGraphNg.class.php [hunk 13]
```diff
@@ -244,7 +246,19 @@ public function __construct($userId)
             $this->rrdCachedOptions[$row['config_key']] = $row['config_value'];
         }
     }
-    
+
+    /**
+     * Tell it will be a multiple services graph
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
     * Set if a graph has multiple services
```

File: www/class/centreonGraphNg.class.php [hunk 14]
```diff
@@ -596,7 +620,58 @@ public function addServiceMetrics($hostId, $serviceId)
...
+    /**
+     * Add metrics for a service
+     *
+     * @param int   $hostId
+     * @param int   $serviceId
+     * @param mixed $metricsSelected
+     *
+     * @return void
+     */
+    public function addServiceCustomMetrics($hostId, $serviceId, $metricsSelected)
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
    public function addServiceCustomMetrics($hostId, $serviceId, $metricsSelected): void
```

File: www/class/centreonGraphNg.class.php [hunk 15]
```diff
@@ -596,7 +620,58 @@ public function addServiceMetrics($hostId, $serviceId)
...
+        $stmt->bindParam(':service_id', $serviceId, PDO::PARAM_INT);
+        $stmt->execute();
+        $metrics = $stmt->fetchAll(PDO::FETCH_ASSOC);
+        foreach ($metrics as $metric) {
+            $indexId = $metric['index_id'];
+            $this->addIndexId($metric['index_id']);
+            if (isset($metricsSelected[$metric['metric_id']])) {
+                $this->addRealMetric($metric);
+            } else {
+                $this->addRealMetric($metric, 1);
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
                // this metric will be hidden
                $this->addRealMetric($metric, 1);
```

File: www/class/centreonGraphNg.class.php [hunk 16]
```diff
@@ -1060,9 +1155,10 @@ public function getJsonStream()
             null,
             null
         );
+        $this->extraDatas['multiple_services'] = $this->multipleServices;
         $this->graphData = array(
             'global' => $this->extraDatas,
-            'metrics' => array(),
+            'metrics' => array()
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            'metrics' => []
```

File: www/class/centreonGraphNg.class.php [hunk 17]
```diff
@@ -1125,19 +1221,20 @@ public function getOVDColor($indexId, $metricId)
     {
         if (is_null($this->colorCache)) {
             $this->colorCache = array();
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
            $this->colorCache = [];
```

File: www/class/centreonGraphNg.class.php [hunk 18]
```diff
@@ -1192,8 +1289,9 @@ public function getRandomWebColor()
             '#ff33ff', '#ff6600', '#ff6633', '#ff6666', '#ff6699', '#ff66cc',
             '#ff66ff', '#ff9900', '#ff9933', '#ff9966', '#ff9999', '#ff99cc',
             '#ff99ff', '#ffcc00', '#ffcc33', '#ffcc66', '#ffcc99', '#ffcccc',
-            '#ffccff');
-            return $webSafeColors[rand(0, sizeof($webSafeColors)-1)];
+            '#ffccff'
+        );
+        return $webSafeColors[rand(0, sizeof($webSafeColors)-1)];
```
Comment (sc979) [inline www/class/centreonGraphNg.class.php]:
```suggestion
        return $webSafeColors[rand(0, sizeof($webSafeColors) - 1)];
```

File: www/api/class/centreon_metric.class.php [hunk 19]
```diff
@@ -313,7 +633,7 @@ public function getStatusByService()
             $query = 'SELECT `value` FROM `options` WHERE `key` = "display_comment_chart"';
             $res = $this->pearDB->query($query);
             $row = $res->fetch();
-            if (false === is_null($row) && $row['value'] === '1') {
+            if (!is_null($row) && $row['value'] === '1') {
```
Comment (loiclau) [inline www/api/class/centreon_metric.class.php]: if 'key' doesn't exist, the fetch method return false, not null.

File: www/api/class/centreon_metric.class.php [hunk 20]
```diff
@@ -437,9 +755,9 @@ protected function serviceDatasNg($id, $metric = null)
         $query = 'SELECT `value` FROM `options` WHERE `key` = "display_downtime_chart"';
 
         $res = $this->pearDB->query($query);
-        
+
         $row = $res->fetch();
-        if (false === is_null($row) && $row['value'] === '1') {
+        if (!is_null($row) && $row['value'] === '1') {
```
Comment (loiclau) [inline www/api/class/centreon_metric.class.php]: if 'key' doesn't exist, the fetch method return false, not null.

**Pull request merged**